scripts/test_all.py

```diff
@@ -0,0 +1,45 @@
+import os
+import csv
+from argparse import ArgumentParser
+from .preprocess import preprocess
+from .test_single import test_single
+from .load_tools import setup_psiminer, setup_code2seq
+
+
+def test_all(dataset_path: str, model_path: str, results_path: str):
+    """Evaluate and store all"""
+
+    project_names = os.listdir(dataset_path)
+    for project_name in project_names:
+        preprocess(os.path.join(dataset_path, project_name))
+
+    project_names = os.listdir("datasets")
+    result_file = os.path.join(results_path, "results.csv")
+    header = ["Project", "F1", "Precision", "Recall", "Loss"]
+    with open(result_file, "w") as f:
+        writer = csv.DictWriter(f, fieldnames=header)
+        writer.writeheader()
+        for project_name in project_names:
+            print(project_name)
+            try:
+                metrics = test_single(model_path, os.path.join("datasets", project_name))
+            except:
+                metrics = [-1, -1, -1, -1]
+            row = {"Project": project_name}
+            for i in range(1, len(header)):
+                row[header[i]] = metrics[i - 1]
+            writer.writerow(row)
+        f.close()
```

Comment:
The point of using “with” block is that it will call close() no matter what happens inside the block (e.g., error). So you don’t need to close the file manually.